text/0000-contingency-plan.md

```diff
@@ -0,0 +1,425 @@
+and soundness problems?** The current text is vague on what precisely
+constitutes a compiler bug and soundness change. It may be worth
+defining more precisely, though likely this would be best done as part
+of writing up a more thorough (and authoritative) Rust reference
+manual.
+
+**Should we add a mechanism for "escaping" keywords?"** We may need a
+mechanism for escaping keywords in the future. Imagine you have a
+public function named `foo`, and we add a keyword `foo`. Now, if you
+opt in to the newer version of Rust, your function declaration is
+illegal: but if you rename the function `foo`, you are making a
+breaking change for your clients, which you may not wish to do. If we
+had an escaping mechanism, you would probably still want to deprecate
+`foo` in favor of a new function `bar` (since typing `foo` would be
+awkward), but it could still exist.
+
+**Should we add a mechanism for skipping over new syntax?** The
+current `#[cfg]` mechanism is applied *after* parsing. This implies
+that if we add new syntax, crates which employ that new syntax will
+not be parsable by older compilers, even if the modules that depend on
+that new syntax are disabled via `#[cfg]` directives. It may be useful
+to add some mechanism for informing the parser that it should skip
+over sections of the input (presumably based on token trees). One
+approach to this might just be modifying the existing `#[cfg]`
+directives so that they are applied during parsing rather than as a
+post-pass.
+
+**What precisely constitutes "small" impact?** This RFC does not
+attempt to define when the impact of a patch is "small" or "not
+small". We will have to develop guidelines over time based on
+precedent. One of the big unknowns is how indicative the breakage we
+observe on `crates.io` will be of the total breakage that will occur:
+it is certainly possible that all crates on `crates.io` work fine, but
+the change still breaks a large body of code we do not have access to.
+
+**Should deprecation due to unsoundness have a special lint?** We may
+not want to use the same deprecation lint for unsoundness that we use
+for everything else.
+
+**What attribute should we use to "opt out" of soundness changes?**
+The section on breaking changes indicated that it may sometimes be
+appropriate to includ an "opt out" that people can use to temporarily
+revert to older, unsound type rules, but did not specify precisely
+what that opt-out should look like. Ideally, we would identify a
+specific attribute in advance that will be used for such purposes.  In
+the past, we have simply created ad-hoc attributes (e.g.,
+`#[old_orphan_check]`), but because custom attributes are forbidden by
+stable Rust, this has the unfortunate side-effect of meaning that code
+which opts out of the newer rules cannot be compiled on older
+compilers (even though it's using the older type system rules). If we
+introduce an attribute in advance we will not have this problem.
```

Comment:
We've talked about scoped attributes before, it seems like we could take advantage of that here. It does have the backwards compatibility issue, but only once, not for every attribute.